aip/general/0213.md

```diff
@@ -9,35 +9,66 @@ placement:
 
 # Common components
 
-In general, API reviewers encourage API producers to keep their APIs mostly
-self-contained, except for a relatively small set of common protos which are
-safe to import (e.g. [`google.protobuf.Timestamp`][timestamp]). This is for
-good reason: APIs generally need to be able to move forward independently of
-one another, and mutual dependencies can cause downstream APIs to be forced
-into taking major version changes or even lead to dependency conflicts.
+As specified in [AIP-215][], APIs must be self-contained except for "common
+protos" which are intended for use by multiple APIs.
 
-However, there are also cases where common structures are valuable, especially
-where a concept is well-known and it is sufficiently clear that it will not
-change. Common protos serve this use case.
+There are two kinds of common protos:
 
-## Guidance
-
-The public representation of APIs **should** be self-contained, meaning that
-all protos used by the API originate in the same proto `package`, except for
-common protos, which **may** be used freely in any API.
+- Organization-specific common protos, covering organization-specific concepts
+  such as a "viewport" in Maps.
+- Global common protos which are not domain-specific, such as "timestamp" or
+  "postal address".
 
-APIs **must not** define a set of API-specific common protos which live outside
-of its versioning structure. This prevents independent movement of particular
-versions and also causes problems for client libraries in many languages that
-compile the proto messages into classes.
+Where it is safe to share a single representation across multiple APIs,
+common protos can make it easier for clients to interact with those APIs.
+Concept-specific client code can be written once, and messages can be used from
+the response of one API in the request of another without clunky copying, for
+example.
 
-APIs **should not** directly depend on protos defined in other APIs. Instead,
-they should copy and paste the applicable messages into their own API. When
-doing so, APIs **should** keep the field names and numbers the same.
+This benefit comes with significant restrictions and limitations, however,
+and should not be attempted lightly.
 
-## Existing common protos
+## Guidance
 
-The common protos, which public-facing protos for an API **may** safely import,
+- Organization-wide common proto packages **must** end with `.type`, e.g.
+  `google.geo.type` or `google.shopping.type`.
+- Organizations **must** consult the API design team before creating a new
+  organization-wide common proto package.
+- Organizations **must not** define generic protos in organization-specific
+  common protos, instead preferring global common protos.
+- Common protos **must not** be "moved" from an organization-specific common
```

Comment:
what if what was originally thought to be an organization-specific concept ends up being more globally applicable? e.g. other organizations would like to use it. Would each organization duplicate the original organization-specific proto?

Reply:
If the type "graduates" to google.type, and remains wire-compatible (serialized proto and JSON), what's the problem with switching to the new copy of the type in google.type?

Reply:
It's a breaking change for API libraries in statically-typed languages, because the type would be different. Removing it from the org-specific common library would also be a breaking change, as removing public types always is.

The latter breaking change is the big problem, as a new major version of the org-specific common library requires a new major version of all API libraries depending on it.

Basically there are ways this could just about be done, but it's a *lot* of work - much better to really take the time to work out global/org-specific first. (The DateRange one earlier is a good example of a gray area... we *may* have got that wrong, but it's unclear. It's not obviously org-specific in that date ranges themselves are quite general - but the detailed requirements of the type around precision, inclusivity etc are org-specific.)

Reply:
If I have an API client that expects `google.shoe.type.ShoeSize` and the server is updated to use a graduated `google.type.ShoeSize`...why would my client break? 

I agree that a generated GAPIC in a statically typed language would contain a breaking change, but that seems distinct from a breaking change in the API itself.  Client code that isn't updated at all would keep working, right?

Reply:
> I agree that a generated GAPIC in a statically typed language would contain a breaking change when regenerated after the switch to the graduated type

And that's what I'm talking about.

> that seems distinct from a breaking change in the API itself.

We've always included "that would cause a breaking change in client libraries" as a breaking change, and I'd want to keep doing so. AIP-180 explicitly calls that out with 3 kinds of compatibility - this would be compatible with item 2 and 3 there, but not with item 1.

> Client code that isn't updated at all would keep working, right?

Yes, unless the type is used as an "Any". But there's more to compatibility than "existing code against existing libraries keeps working".

Reply:
Yeah fair enough - since AIP-180's item 1 is explicitly includes newer versions of client libraries, that's that I suppose.

Reply:
@toumorokoshi: Are you happy to resolve this now?